src/psyclone/psyir/tools/call_tree_utils.py

```diff
@@ -0,0 +1,415 @@
+
+        :returns: a ReadWriteInfo object with the information about input-
+            and output parameters.
+        :rtype: :py:class:`psyclone.psyir.tools.ReadWriteInfo`
+
+        '''
+        variables_info = VariablesAccessInfo(node_list, options=options)
+        read_write_info = ReadWriteInfo()
+        self.get_input_parameters(read_write_info, node_list, variables_info)
+        self.get_output_parameters(read_write_info, node_list, variables_info)
+        return read_write_info
+
+    # -------------------------------------------------------------------------
+    def get_non_local_read_write_info(self, node_list, read_write_info):
+        '''Returns the information about non-local variables that are read
+        or written.
+
+        '''
+
+        # First collect all non-local symbols from the kernels called. They
+        # are collected in the todo list. This list will initially contain
+        # unknown accesses, since at this stage we cannot always differentiate
+        # between function calls and array accesses. In the resolve step
+        # that is following the corresponding modules will be queried and
+        # the right accesses (functions or variables) will be used.
+        todo = []
+        mod_manager = ModuleManager.get()
+        for node in node_list:
+            for kernel in node.walk(Kern):
+                if isinstance(kernel, BuiltIn):
+                    # Builtins don't have non-local accesses
+                    continue
+
+                # Get the non-local access information from the kernel
+                # by querying the module that contains the kernel:
+                try:
+                    mod_info = mod_manager.get_module_info(kernel.module_name)
+                except FileNotFoundError:
+                    print(f"Could not find module '{kernel.module_name}' - "
+                          f"ignored.")
+                    continue
+
+                all_possible_routines = mod_info.resolve_routine(kernel.name)
+                for routine_name in all_possible_routines:
+                    psyir = mod_info.get_psyir(routine_name)
+                    todo.extend(self.get_non_local_symbols(psyir))
+        return self._resolve_calls_and_unknowns(todo, read_write_info)
+
+    # -------------------------------------------------------------------------
+    def _resolve_calls_and_unknowns(self, todo, read_write_info):
+        '''This function updates the list of non-local symbols by:
+        1. replacing all subroutine calls with the list of their corresponding
+            non-local symbols.
+        2. Resolving unknown types to be either function calls (which then
+            get resolved as above), or variables.
+        This is done recursively until only variable accesses remain.
+
+        The actual non-local accesses will then be added to the ReadWriteInfo
+        object.
+
+        :param todo: the information about symbol type, module_name,
```

Comment:
I think this needs a less confusing name - perhaps `outstanding_nonlocals` or something?

Reply:
Done.

Reply:
I meant s/todo/outstanding_nonlocals/ :-) I was quite happy with the routine name - please could you revert it.